fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateOutJoin.java

```diff
@@ -0,0 +1,109 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.rules.rewrite.logical;
+
+import org.apache.doris.nereids.rules.Rule;
+import org.apache.doris.nereids.rules.RuleType;
+import org.apache.doris.nereids.rules.expression.rewrite.ExpressionRewriteContext;
+import org.apache.doris.nereids.rules.expression.rewrite.rules.FoldConstantRule;
+import org.apache.doris.nereids.rules.rewrite.OneRewriteRuleFactory;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.Slot;
+import org.apache.doris.nereids.trees.expressions.literal.BooleanLiteral;
+import org.apache.doris.nereids.trees.expressions.literal.Literal;
+import org.apache.doris.nereids.trees.plans.GroupPlan;
+import org.apache.doris.nereids.trees.plans.JoinType;
+import org.apache.doris.nereids.trees.plans.logical.LogicalJoin;
+import org.apache.doris.nereids.util.ExpressionUtils;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Eliminate out join.
+ */
+public class EliminateOutJoin extends OneRewriteRuleFactory {
+
+    @Override
+    public Rule build() {
+        return logicalFilter(logicalJoin())
+                .when(filter -> filter.child().getJoinType().isOuterJoin())
+                .then(filter -> {
+                    LogicalJoin<GroupPlan, GroupPlan> join = filter.child();
+                    List<Expression> leftPredicates = Lists.newArrayList();
+                    List<Expression> rightPredicates = Lists.newArrayList();
+                    for (Expression predicate : ExpressionUtils.extractConjunction(filter.getPredicates())) {
+                        if (allCoveredBy(predicate, join.left().getOutputSet())) {
+                            leftPredicates.add(predicate);
+                        }
+                        if (allCoveredBy(predicate, join.right().getOutputSet())) {
+                            rightPredicates.add(predicate);
+                        }
+                    }
```

Comment:
nit: the current code abstraction level is not uniform, add a layer of abstraction for the detail of compute.

```java
List<Expression> leftPredicates = ExpressionUtils.extractCoveredConjunction(filter.getPredicates(), join.left().getOutputSet());
List<Expression> rightPredicates = ExpressionUtils.extractCoveredConjunction(filter.getPredicates(), join.right().getOutputSet());
```

keep in mind: rule should display the core logical firstly, and hide the compute detail in other methods or util class.